ambry-network/src/main/java/com.github.ambry.network/SSLBlockingChannel.java

```diff
@@ -51,75 +48,42 @@ public SSLBlockingChannel(String host, int port, MetricRegistry registry, int re
   }
 
   @Override
-  public void connect() throws IOException {
-    synchronized (lock) {
-      if (!connected) {
-        Socket socket = new Socket();
-        socket.setSoTimeout(readTimeoutMs);
-        socket.setKeepAlive(true);
-        socket.setTcpNoDelay(true);
-        if (readBufferSize > 0) {
-          socket.setReceiveBufferSize(readBufferSize);
-        }
-        if (writeBufferSize > 0) {
-          socket.setSendBufferSize(writeBufferSize);
-        }
-        socket.connect(new InetSocketAddress(host, port), connectTimeoutMs);
-        sslSocket = (SSLSocket) sslSocketFactory.createSocket(socket, host, port, true);
+  public Socket createSocket() throws IOException {
+    Socket tcpSocket = super.createSocket();
 
-        ArrayList<String> protocolsList = Utils.splitString(sslConfig.sslEnabledProtocols, ",");
-        if (!protocolsList.isEmpty()) {
-          String[] enabledProtocols = protocolsList.toArray(new String[protocolsList.size()]);
-          sslSocket.setEnabledProtocols(enabledProtocols);
-        }
+    SSLSocket sslSocket = null;
 
-        ArrayList<String> cipherSuitesList = Utils.splitString(sslConfig.sslCipherSuites, ",");
-        if (!cipherSuitesList.isEmpty()) {
-          String[] cipherSuites = cipherSuitesList.toArray(new String[cipherSuitesList.size()]);
-          sslSocket.setEnabledCipherSuites(cipherSuites);
-        }
+    try {
+      sslSocket = (SSLSocket) sslSocketFactory.createSocket(tcpSocket, host, port, true);
 
-        // handshake in a blocking way
-        try {
-          sslSocket.startHandshake();
-          sslClientHandshakeCount.inc();
-        } catch (IOException e) {
-          sslClientHandshakeErrorCount.inc();
-          throw e;
-        }
-        writeChannel = Channels.newChannel(sslSocket.getOutputStream());
-        readChannel = sslSocket.getInputStream();
-        connected = true;
-        logger.debug(
-            "Created socket with SO_TIMEOUT = {} (requested {}), SO_RCVBUF = {} (requested {}), SO_SNDBUF = {} (requested {})",
-            sslSocket.getSoTimeout(), readTimeoutMs, sslSocket.getReceiveBufferSize(), readBufferSize,
-            sslSocket.getSendBufferSize(), writeBufferSize);
+      ArrayList<String> protocolsList = Utils.splitString(sslConfig.sslEnabledProtocols, ",");
+      if (!protocolsList.isEmpty()) {
+        String[] enabledProtocols = protocolsList.toArray(new String[protocolsList.size()]);
```

Comment:
Why are you going from config string to ArrayList to array of Strings?  String.split() does that in one shot.

Reply:
There is a slight gotcha with `String.split()`. If the string passed to it is "", it will return an array with 1 empty string in it, so the check after will not work. 

I think it would be best if `SSLBlockingChannel` did not need to be given an `SSLConfig` object and parse the configs every time it creates a new connection.  Instead, the `SSLFactory` could provide a method for getting an `SSLSocket` instead of an `SSLContext` object. However, that may be out of the scope of this PR and should be done in a follow up.

Reply:
I agree - lets not make too many changes this time around, as the refactoring alone is quite a significant change.

The string indeed should be split and stored as an array only once, though, and we definitely should make that change sooner rather than later.

Reply:
That's fine.  It was hard to tell how much of the code is new vs moved around since I'm not that familiar with it.  I was just pointing out the deficits that I noticed.